samples/nrf9160/nrf_cloud_rest_device_message/src/main.c

```diff
@@ -0,0 +1,364 @@
+
+#define REST_RX_BUF_SZ 2100
+/* Buffer used for REST calls */
+static char rx_buf[REST_RX_BUF_SZ];
+
+/* nRF Cloud REST context */
+static struct nrf_cloud_rest_context rest_ctx = {
+	.connect_socket = -1,
+	.keep_alive = false,
+	.rx_buf = rx_buf,
+	.rx_buf_len = sizeof(rx_buf),
+	.fragment_size = 0
+};
+
+/* Flag to indicate if the user requested JITP to be performed */
+static bool jitp_requested;
+
+static int set_led(const int led, const int state)
+{
+	int err = dk_set_led(led, state);
+
+	if (err) {
+		LOG_ERR("Failed to set LED %d, error: %d", led, err);
+		return err;
+	}
+	return 0;
+}
+
+static int init_leds(void)
+{
+	int err = dk_leds_init();
+
+	if (err) {
+		LOG_ERR("LED init failed, error: %d", err);
+		return err;
+	}
+
+	err = set_led(LTE_LED_NUM, 0);
+	if (err) {
+		return err;
+	}
+
+	err = set_led(SEND_LED_NUM, 0);
+	if (err) {
+		return err;
+	}
+
+	return err;
+}
+
+static void button_handler(uint32_t button_states, uint32_t has_changed)
+{
+	if (has_changed & button_states & BIT(BTN_NUM - 1)) {
+		LOG_DBG("Button %d pressed", BTN_NUM);
+		k_sem_give(&button_press_sem);
+	}
+}
+
+static int send_message(const char *const msg)
+{
+	int ret = 0;
+
+	/* Turn the SEND LED on for a bit */
+	set_led(SEND_LED_NUM, 1);
+
+	/* Send the message to nRF Cloud */
+	ret = nrf_cloud_rest_send_device_message(&rest_ctx, device_id, msg, false, NULL);
+	if (ret) {
+		LOG_ERR("Failed to send device message via REST: %d", ret);
+	}
+
+	/* Keep that LED on for at least 100ms */
+	k_sleep(K_MSEC(100));
+
+	/* Turn the LED back off */
+	set_led(SEND_LED_NUM, 0);
+
+	return ret;
+}
+
+static void send_message_on_button(void)
+{
+	/* Wait for a button press */
+	(void)k_sem_take(&button_press_sem, K_FOREVER);
+
+	/* Send off a JSON message about the button press */
+	/* This matches the nRF Cloud-defined schema for a "BUTTON" device message */
+	(void)send_message("{\"appId\":\"BUTTON\", \"messageType\":\"DATA\", \"data\":\"1\"}");
+}
+
+static int do_jitp(void)
+{
+	int ret = 0;
+
+	LOG_INF("Performing JITP...");
+
+	/* Turn the SEND LED, indicating in-progress JITP*/
+	set_led(SEND_LED_NUM, 1);
+
+	ret = nrf_cloud_rest_jitp(CONFIG_NRF_CLOUD_SEC_TAG);
+
+	if (ret == 0) {
+		LOG_INF("Waiting 30s for cloud provisioning to complete...");
+		k_sleep(K_SECONDS(30));
+		k_sem_reset(&button_press_sem);
+		LOG_INF("Associate device with nRF Cloud account and press button %d when complete",
+			BTN_NUM);
+		(void)k_sem_take(&button_press_sem, K_FOREVER);
+	} else if (ret == 1) {
+		LOG_INF("Device already provisioned");
+		ret = 0;
+	} else {
+		LOG_ERR("JITP device provisioning failed: %d", ret);
+	}
+
+	/* Turn LED back off */
+	set_led(SEND_LED_NUM, 0);
+
+	return ret;
+}
+
+static int modem_time_wait(void)
+{
+	int err;
+	char time_buf[64];
+
+	LOG_INF("Waiting for modem to acquire network time...");
+
+	do {
+		k_sleep(K_SECONDS(3));
+		err = nrf_modem_at_cmd(time_buf, sizeof(time_buf), "AT%%CCLK?");
+	} while (err != 0);
+
+	if (err) {
```

Comment:
It looks that it's not even possible to get `err != 0` here - the above loop will not break until `nrf_modem_at_cmd()` succeeds. Maybe just move the log inside the loop and skip the return part?

Reply:
Oh my, how silly of me. Good catch.